src/main/kotlin/uk/gov/justice/digital/hmpps/nomisprisonerapi/locations/LocationService.kt

```diff
@@ -49,7 +54,152 @@ class LocationService(
     return LocationIdResponse(
       agencyInternalLocationRepository.save(
         locationDto.toAgencyInternalLocation(locationType, housingUnitType, agency, parent),
-      ).locationId,
+      ).also {
+        telemetryClient.trackEvent(
+          "location-created",
+          mapOf(
+            "locationId" to it.locationId.toString(),
+            "description" to it.description,
+          ),
+          null,
+        )
+      }.locationId,
+    )
+  }
+
+  @Audit
+  fun updateLocation(locationId: Long, locationDto: UpdateLocationRequest) {
+    val location = agencyInternalLocationRepository.findByIdOrNull(locationId)
+      ?: throw NotFoundException("Location with id=$locationId does not exist")
+
+    val internalLocationType =
+      internalLocationTypeRepository.findByIdOrNull(InternalLocationType.pk(locationDto.locationType))
+        ?: throw BadDataException("Location type with id=${locationDto.locationType} does not exist")
+
+    val housingUnitType = locationDto.unitType?.let {
+      housingUnitTypeRepository.findByIdOrNull(HousingUnitType.pk(it))
+        ?: throw BadDataException("Housing unit type with id=${locationDto.unitType} does not exist")
+    }
+
+    val parent = locationDto.parentLocationId?.let {
+      agencyInternalLocationRepository.findByIdOrNull(it)
+        ?: throw BadDataException("Parent location with id=$it does not exist")
+    }
+
+    location.apply {
+      description =
+        location.parentLocation?.let { "${location.parentLocation!!.description}-${locationDto.locationCode}" }
+          ?: locationDto.locationCode
+      locationType = internalLocationType
+      description = locationDto.description
+      userDescription = locationDto.userDescription
+      locationCode = locationDto.locationCode
+      parentLocation = parent
+      listSequence = locationDto.listSequence
+      comment = locationDto.comment
+      unitType = housingUnitType
+    }.also {
+      telemetryClient.trackEvent(
+        "location-updated",
+        mapOf(
+          "locationId" to it.locationId.toString(),
+          "description" to it.description,
+        ),
+        null,
+      )
+    }
+  }
+
+  @Audit
+  fun deactivateLocation(locationId: Long, deactivateRequest: DeactivateRequest) {
+    val location = agencyInternalLocationRepository.findByIdOrNull(locationId)
+      ?: throw NotFoundException("Location with id=$locationId does not exist")
+
+    if (location.deactivateDate != null) {
+      if (location.reactivateDate == null || location.reactivateDate!!.isBefore(location.deactivateDate)) {
+        throw BadDataException("Location with id=$locationId is already deactivated")
+      }
+    }
+
+    location.deactivateDate = LocalDate.now()
+    location.deactivateReason = deactivateRequest.reasonCode?.let {
```

Comment:
am I right in thinking they don't have to give a reason for a deactivate request?  Why is that?

Reply:
Looking at existing data there is usually no reason entered